sherpa/astro/background.py

```diff
@@ -61,20 +63,58 @@ def add_response(session, id, data, model):
         background components.
 
     """
-
+    id = session._fix_id(id)
     # QUS: if this gets used to generate the response for the
     #      background then how does it pick up the correct response
     #      (ie when fit_bkg is used). Or does that get generated
     #      by a different code path?
+    pileup_model = session._pileup_models.get(id)
+    bkg_srcs = session._background_sources.get(id, {})
+    return get_response_for_pha(data, model, bkg_srcs, pileup_model, id)
 
-    resp = session._get_response(id, data)
-    if data.subtracted:
-        return resp(model)
 
-    bkg_srcs = session._background_sources.get(id, {})
-    if len(bkg_srcs) == 0:
-        return resp(model)
+def get_response_for_pha(data, model, bkg_srcs={}, pileup_model=None, id=None):
```

Comment:
I don't like the "id" parameter, but I did not find another way to pass it down from the UI layer to get the correct error messages.

Reply:
Requiring id is a bit of a code smell (we have this in other places, in particular model expressions passed as a string) since we really shouldn't have the upper-level concepts moving down to this code. That's not to say this particular change is bad (I haven't looked beyond this response) but we should see what would be needed to remove this argument.

Reply:
I will note though that `add_response` takes an `id` value so the question might be whether we expect this routine to be used outside of the session class?

Reply:
If by "this" you mean the new "get_response_for_pha", then the answer is yes. I need to for the faking, that's why I moved it out of the session class. Note that id does not do anything except appearing in the string of the warning that's emitted when a pile-up model is present. So the code itself is "clean" - a good indication that this should indeed be a generic function and not part of the session class. I considered rewording the warning to something so generic that it does not need the id in the string, but, from the UI layer, it's actually very useful to see which dataset caused the warning. Another option would be to pass the warning back up, catch it, and replace it with the text we want in the session class. But that's more complex than it's worth.